Improve Verify Error Responses

ChaosInTheCRD opened the pull request.

colek42 commented on policy/policy.go:
```diff
@@ -1,4 +1,4 @@
-// Copyright 2021 The Witness Contributors
+// Copyright 2021 The Witness Contributor
```
typo

ChaosInTheCRD replied:
```suggestion
// Copyright 2021 The Witness Contributors
```

colek42 commented on policy/step.go:
```diff
@@ -55,10 +56,36 @@ type RegoPolicy struct {
 	Rejected []RejectedCollection
 }
 
+func (r StepResult) Analyze() bool {
```
Please add a comment here -- what does Analyze do?

colek42 commented:
```
➜  policy-service git:(main) ✗ ./test.sh
+ SHA=7ca1e69b7f80e02346819160df24503484ecf268c09a19528ab5a4cb12944af5
+ go run . run --archivista-url https://archivista.testifysec.io/ --subject-digest 7ca1e69b7f80e02346819160df24503484ecf268c09a19528ab5a4cb12944af5 --kms-url=awskms:///arn:aws:kms:us-east-2:184822323247:key/a5b221b1-0670-4c6d-8760-135cb29a0ad5 --sticky-keys-file=sticky.yaml --tsa=https://freetsa.org/files/cacert.pem
+ jq -r
+ witness sign -k policykey.pem -o auto-policy-signed.json -f auto-policy.json
+ witness verify -p auto-policy-signed.json -k policypub.pem -s -l debug --enable-archivista -s 7ca1e69b7f80e02346819160df24503484ecf268c09a19528ab5a4cb12944af5
INFO    policy signature verified
INFO    Verification succeeded
INFO    Evidence:
INFO    0: acaf3d0bb40b715a21e42320517276dae28e999b1f80f028ed7095fb2ae7275f
INFO    1: b43f217cc65b92ef2b1c0ee3041807a7867e475cac080592779fd2de4a3097ac
INFO    2: bb60e7f30f990fccc991649b8057ac385278666d894d04941b1e36b63e340c6a
INFO    3: 2a6eb1ba5f048799079fe941ac5929a557ac8b1667fdfaff6e59e5d85aafd0bd
INFO    4: b222369b43821e349c1f4cc9051c9875b35c50b431af6b00f3dedb08d4aa8dce
INFO    5: 24260cb11925061c988f134a7cfeb21ab16b02cfe51f091d37c96f37f221970c
INFO    6: f00e5177ee0fccd9b28e2299c8c1de3a5f5de96506e6cfba46ce1ca9b408eb4a
INFO    7: 62e2530eb74c2dead4b20872f67e9d0c1211663ad0904f64c45058265773484c
INFO    8: 9327c17ef676d7b35904faaa03065b61e6b3fc91caae0710c0cd4ee54fc196b2
INFO    9: e9be26c45b7311c02291924e25bd500ba9422632e3ccc1793b556956f2a68520
INFO    10: dcb4df0f81838f8917fce371de6bd6f11ec4413f044c32693b76f9027118dd9c
INFO    11: 2d3f4a544a61f0ee2cbd6426e364dd05d0b77c0ae58568fa31e721be7ecc160a
+ witness2 verify -p auto-policy-signed.json -k policypub.pem -s -l debug --enable-archivista -s 7ca1e69b7f80e02346819160df24503484ecf268c09a19528ab5a4cb12944af5
INFO    policy signature verified
ERROR   Verification failed
ERROR   Evidence:
ERROR   Step: docker-build
ERROR   Step: release
ERROR   Step: cve-scan
ERROR   Step: secret-scan
ERROR   Step: test
ERROR   Step: build
ERROR   Verification Failure: Reason: failed to verify artifacts for step build: no passed collections present
ERROR   Step: fmt
ERROR   Step: generate-sbom
ERROR   Step: vet
ERROR   Step: checkout
ERROR   Step: sast
ERROR   Step: lint
```

This is failing on one of the policies I am using for the GitLab demo

ChaosInTheCRD commented:
> ```
> ➜  policy-service git:(main) ✗ ./test.sh
> + SHA=7ca1e69b7f80e02346819160df24503484ecf268c09a19528ab5a4cb12944af5
> + go run . run --archivista-url https://archivista.testifysec.io/ --subject-digest 7ca1e69b7f80e02346819160df24503484ecf268c09a19528ab5a4cb12944af5 --kms-url=awskms:///arn:aws:kms:us-east-2:184822323247:key/a5b221b1-0670-4c6d-8760-135cb29a0ad5 --sticky-keys-file=sticky.yaml --tsa=https://freetsa.org/files/cacert.pem
> + jq -r
> + witness sign -k policykey.pem -o auto-policy-signed.json -f auto-policy.json
> + witness verify -p auto-policy-signed.json -k policypub.pem -s -l debug --enable-archivista -s 7ca1e69b7f80e02346819160df24503484ecf268c09a19528ab5a4cb12944af5
> INFO    policy signature verified
> INFO    Verification succeeded
> INFO    Evidence:
> INFO    0: acaf3d0bb40b715a21e42320517276dae28e999b1f80f028ed7095fb2ae7275f
> INFO    1: b43f217cc65b92ef2b1c0ee3041807a7867e475cac080592779fd2de4a3097ac
> INFO    2: bb60e7f30f990fccc991649b8057ac385278666d894d04941b1e36b63e340c6a
> INFO    3: 2a6eb1ba5f048799079fe941ac5929a557ac8b1667fdfaff6e59e5d85aafd0bd
> INFO    4: b222369b43821e349c1f4cc9051c9875b35c50b431af6b00f3dedb08d4aa8dce
> INFO    5: 24260cb11925061c988f134a7cfeb21ab16b02cfe51f091d37c96f37f221970c
> INFO    6: f00e5177ee0fccd9b28e2299c8c1de3a5f5de96506e6cfba46ce1ca9b408eb4a
> INFO    7: 62e2530eb74c2dead4b20872f67e9d0c1211663ad0904f64c45058265773484c
> INFO    8: 9327c17ef676d7b35904faaa03065b61e6b3fc91caae0710c0cd4ee54fc196b2
> INFO    9: e9be26c45b7311c02291924e25bd500ba9422632e3ccc1793b556956f2a68520
> INFO    10: dcb4df0f81838f8917fce371de6bd6f11ec4413f044c32693b76f9027118dd9c
> INFO    11: 2d3f4a544a61f0ee2cbd6426e364dd05d0b77c0ae58568fa31e721be7ecc160a
> + witness2 verify -p auto-policy-signed.json -k policypub.pem -s -l debug --enable-archivista -s 7ca1e69b7f80e02346819160df24503484ecf268c09a19528ab5a4cb12944af5
> INFO    policy signature verified
> ERROR   Verification failed
> ERROR   Evidence:
> ERROR   Step: docker-build
> ERROR   Step: release
> ERROR   Step: cve-scan
> ERROR   Step: secret-scan
> ERROR   Step: test
> ERROR   Step: build
> ERROR   Verification Failure: Reason: failed to verify artifacts for step build: no passed collections present
> ERROR   Step: fmt
> ERROR   Step: generate-sbom
> ERROR   Step: vet
> ERROR   Step: checkout
> ERROR   Step: sast
> ERROR   Step: lint
> ```
> 
> This is failing on one of the policies I am using for the GitLab demo. It is also interesting that it fails on a different step each time. I'll add this as a text fixture

I think i've figured out why this is happening. Bear with me while I try and resolve.

ChaosInTheCRD commented:
> ```
> ➜  policy-service git:(main) ✗ ./test.sh
> + SHA=7ca1e69b7f80e02346819160df24503484ecf268c09a19528ab5a4cb12944af5
> + go run . run --archivista-url https://archivista.testifysec.io/ --subject-digest 7ca1e69b7f80e02346819160df24503484ecf268c09a19528ab5a4cb12944af5 --kms-url=awskms:///arn:aws:kms:us-east-2:184822323247:key/a5b221b1-0670-4c6d-8760-135cb29a0ad5 --sticky-keys-file=sticky.yaml --tsa=https://freetsa.org/files/cacert.pem
> + jq -r
> + witness sign -k policykey.pem -o auto-policy-signed.json -f auto-policy.json
> + witness verify -p auto-policy-signed.json -k policypub.pem -s -l debug --enable-archivista -s 7ca1e69b7f80e02346819160df24503484ecf268c09a19528ab5a4cb12944af5
> INFO    policy signature verified
> INFO    Verification succeeded
> INFO    Evidence:
> INFO    0: acaf3d0bb40b715a21e42320517276dae28e999b1f80f028ed7095fb2ae7275f
> INFO    1: b43f217cc65b92ef2b1c0ee3041807a7867e475cac080592779fd2de4a3097ac
> INFO    2: bb60e7f30f990fccc991649b8057ac385278666d894d04941b1e36b63e340c6a
> INFO    3: 2a6eb1ba5f048799079fe941ac5929a557ac8b1667fdfaff6e59e5d85aafd0bd
> INFO    4: b222369b43821e349c1f4cc9051c9875b35c50b431af6b00f3dedb08d4aa8dce
> INFO    5: 24260cb11925061c988f134a7cfeb21ab16b02cfe51f091d37c96f37f221970c
> INFO    6: f00e5177ee0fccd9b28e2299c8c1de3a5f5de96506e6cfba46ce1ca9b408eb4a
> INFO    7: 62e2530eb74c2dead4b20872f67e9d0c1211663ad0904f64c45058265773484c
> INFO    8: 9327c17ef676d7b35904faaa03065b61e6b3fc91caae0710c0cd4ee54fc196b2
> INFO    9: e9be26c45b7311c02291924e25bd500ba9422632e3ccc1793b556956f2a68520
> INFO    10: dcb4df0f81838f8917fce371de6bd6f11ec4413f044c32693b76f9027118dd9c
> INFO    11: 2d3f4a544a61f0ee2cbd6426e364dd05d0b77c0ae58568fa31e721be7ecc160a
> + witness2 verify -p auto-policy-signed.json -k policypub.pem -s -l debug --enable-archivista -s 7ca1e69b7f80e02346819160df24503484ecf268c09a19528ab5a4cb12944af5
> INFO    policy signature verified
> ERROR   Verification failed
> ERROR   Evidence:
> ERROR   Step: docker-build
> ERROR   Step: release
> ERROR   Step: cve-scan
> ERROR   Step: secret-scan
> ERROR   Step: test
> ERROR   Step: build
> ERROR   Verification Failure: Reason: failed to verify artifacts for step build: no passed collections present
> ERROR   Step: fmt
> ERROR   Step: generate-sbom
> ERROR   Step: vet
> ERROR   Step: checkout
> ERROR   Step: sast
> ERROR   Step: lint
> ```
> 
> This is failing on one of the policies I am using for the GitLab demo. It is also interesting that it fails on a different step each time. I'll add this as a text fixture

I think the issue is that the depthened search creates a situation where the `stepResult` is getting overwritten. Therefore I have added [logic](https://github.com/in-toto/go-witness/pull/210/files#diff-82bdb5b5421b9da021e9c0af06d41fa22ff7ac433bb0c7cbef393c60be371c36R230-R240) so the results get merged together, and moved [verifyArtifacts](https://github.com/in-toto/go-witness/pull/210/files#diff-82bdb5b5421b9da021e9c0af06d41fa22ff7ac433bb0c7cbef393c60be371c36R251-R253) outside of the search loop

jkjell commented on verify.go:
```diff
@@ -104,7 +104,7 @@ func VerifyWithPolicyCertConstraints(commonName string, dnsNames []string, email
 // Verify verifies a set of attestations against a provided policy. The set of attestations that satisfy the policy will be returned
 // if verifiation is successful.
-func Verify(ctx context.Context, policyEnvelope dsse.Envelope, policyVerifiers []cryptoutil.Verifier, opts ...VerifyOption) (map[string][]source.VerifiedCollection, error) {
+func Verify(ctx context.Context, policyEnvelope dsse.Envelope, policyVerifiers []cryptoutil.Verifier, opts ...VerifyOption) (bool, map[string]policy.StepResult, error) {
```
This is a breaking API change. We'll need to deprecate the original function signature.

ChaosInTheCRD replied:
Good point.

jkjell commented on verify.go:
```diff
@@ -171,12 +171,21 @@ func Verify(ctx context.Context, policyEnvelope dsse.Envelope, policyVerifiers [
+		return false, nil, fmt.Errorf("error encountered during policy verification: %w", err)
+	}
+
+	pass := true
+	for _, result := range results {
+		p := result.Analyze()
+		if !p {
+			pass = false
+		}
 	}
```
Can we do this at the end of pol.Verify? it's a bit confusing for the Verify function to return without an error if the policy is actually failing.

ChaosInTheCRD replied:
So I'm glad that this has come up, it's something I wasn't 100% sure on.

The intention here was to change the usage of `err` so it is reserved for functional errors / issues during the `Verify` flow (e.g., `failed to open file`). This is why the `bool` was introduced as a clear way of signifying if the policy verification failed or passed. I was however also concerned that this would end up confusing for users.

jkjell reviewed (commented):
I think we need to retitle this PR... it's way more than just improving error responses. It's a refactor of a lot of the policy verification workflow.

ChaosInTheCRD commented on policy/policy.go:
```diff
@@ -214,94 +214,134 @@ func (p Policy) Verify(ctx context.Context, opts ...VerifyOption) (map[string][]
+				}
+			}
+
+			for _, coll := range stepResult.Passed {
```
So in the way that the `stepResult` works, each collection that was found is split into `Passed` and `Rejected` collections. If there are no passing collections, the policy fail by definition. The `collections.Rejected[n].Reason`'s will be exposed to the user as error logs. So basically, if there are no passing results for a step, they will know it and know why.

ChaosInTheCRD commented on policy/policy.go:
```diff
@@ -214,94 +214,134 @@ func (p Policy) Verify(ctx context.Context, opts ...VerifyOption) (map[string][]
 // materials of the original step.  This ensures file integrity between each step.
-func (p Policy) verifyArtifacts(collectionsByStep map[string][]source.VerifiedCollection) (map[string][]source.VerifiedCollection, error) {
-	acceptedByStep := make(map[string][]source.VerifiedCollection)
+func (p Policy) verifyArtifacts(resultsByStep map[string]StepResult) (map[string]StepResult, error) {
```
done 😄

ChaosInTheCRD commented on policy/policy.go:
```diff
@@ -214,94 +214,134 @@ func (p Policy) Verify(ctx context.Context, opts ...VerifyOption) (map[string][]
+				result.Rejected = append(result.Rejected, RejectedCollection{Reason: fmt.Errorf("failed to verify artifacts for step %s: no passed collections present", step.Name)})
+				resultsByStep[step.Name] = result
 			}
+			return resultsByStep, nil
```
Yeah my apologies on this one, very unclear. I have now changed things so it continues if there are no passing results.

ChaosInTheCRD replied:
actually gone an backtracked a bit. We want to handle an error in the case of `!ok` because that implies that there was a problem fetching a step result (which should definitely be there)

jkjell reviewed (approved):
Thanks for all the changes!
